### PR TITLE
fix(gs-cache): cleanup cache usage, don't try to save key_db_entry no…

### PR DESCRIPTION
**Pull request by Avantol13**
### Bug Fixes
- Fix intermittent issue with db cache containing info that can't be loaded into JSON

### Dependency updates
- Update poetry install in travis.yml due to https://python-poetry.org/blog/announcing-poetry-1.2.0a1/#deprecation-of-the-get-poetrypy-script

### CODE REVIEW

File: fence/blueprints/data/indexd.py [hunk 1]
```diff
@@ -1251,18 +1247,18 @@ def _generate_google_storage_signed_url(
...
                 private_key,
-                key_db_entry,
                 key_db_entry.expires,
             )
 
             db_entry = {}
             db_entry["gcp_proxy_group_id"] = proxy_group_id
-            db_entry["gcp_private_key"] = json.dumps(str(private_key))
-            db_entry["gcp_key_db_entry"] = str(key_db_entry)
+            db_entry["gcp_private_key"] = str(private_key)
```
Comment (Avantol13) [inline fence/blueprints/data/indexd.py:1255]: why don't we just json.dumps this and then json.loads should work without the need to do that string manipulation?

```
>>> import json
>>> test = {"foo": "bar"}
>>> db_entry = json.dumps(test)
>>> db_entry
'{"foo": "bar"}'
>>> json.loads(db_entry)
{'foo': 'bar'}
```

File: fence/blueprints/data/indexd.py [hunk 2]
```diff
@@ -1214,23 +1214,19 @@ def _generate_google_storage_signed_url(
                     .first()
                 )
                 if cache and cache.expires_at > expiration_time:
-                    rv = (
-                        json.loads(cache.gcp_private_key),
-                        json.loads(cache.gcp_key_db_entry),
-                        cache.expires_at,
+                    private_key = json.loads(
+                        str(cache.gcp_private_key).replace("'", '"')
```
Comment (Avantol13) [inline fence/blueprints/data/indexd.py:1218]: I'm not sure I understand why this is necessary now. See next comment

File: tests/data/test_indexed_file.py [hunk 3]
```diff
@@ -441,9 +441,7 @@ def test_internal_get_gs_signed_url_cache_new_key_if_old_key_expired(
             ):
                 indexed_file = IndexedFile(file_id="some id")
                 google_object = GoogleStorageIndexedFileLocation("gs://some/location")
-                keydbentry = UserGoogleAccountToProxyGroup()
-                keydbentry.expires = 10
-                google_object._assume_role_cache_gs = {"1": ("key", keydbentry, 10)}
+                google_object._assume_role_cache_gs = {"1": ("key", 10)}
```
Comment (Avantol13) [inline tests/data/test_indexed_file.py:444]: we could probably add some more unit tests here to regression test this? What do you think , I'm not sure we have to have integration tests

File: fence/blueprints/data/indexd.py (hunk 1 above)
Comment (BinamB) [reply fence/blueprints/data/indexd.py]: okay yes, done. Tested it out and its working

File: tests/data/test_indexed_file.py [hunk 4]
```diff
@@ -461,6 +459,108 @@ def test_internal_get_gs_signed_url_cache_new_key_if_old_key_expired(
...
+            return_value=("sa_private_key"),
+        ):
+            with mock.patch.object(
+                cirrus.google_cloud.utils,
+                "get_signed_url",
+                return_value="https://cloud.google.com/compute/url",
+            ):
+                indexed_file = IndexedFile(file_id="some id")
+                google_object = GoogleStorageIndexedFileLocation("gs://some/location")
+                google_object._assume_role_cache_gs = {"1": ("key", 10)}
```
Comment (Avantol13) [inline tests/data/test_indexed_file.py:518]: maybe we should use a dictionary as the key to better simulate the real world data (and hopefully catch any issues with the JSON parsing)

File: tests/data/test_indexed_file.py [hunk 5]
```diff
@@ -461,6 +459,108 @@ def test_internal_get_gs_signed_url_cache_new_key_if_old_key_expired(
...
+                    expires_in=0,
+                    user_id=1,
+                    username="some user",
+                    r_pays_project=None,
+                )
+
+                after_cache = db_session.query(AssumeRoleCacheGCP).first()
+
+                assert after_cache
+                assert before_cache != after_cache
```
Comment (Avantol13) [inline tests/data/test_indexed_file.py:535]: Why do we expect these to be different? Also I think we should test the actual contents of the cache, make sure it's a dictionary or JSON (whatever we actually expect in the code)

File: tests/data/test_indexed_file.py [hunk 6]
```diff
@@ -461,6 +459,108 @@ def test_internal_get_gs_signed_url_cache_new_key_if_old_key_expired(
...
+    """
+    Test fence.blueprints.data.indexd.GoogleStorageIndexedFileLocation._generate_google_storage_signed_url does not use cached key if its expired
+
+    create presigned url
+        set cache in db
+    clear cache
+    create presigned url again
+        make sure cache is set correctly
+    """
+    # db_session.add(
```
Comment (Avantol13) [inline tests/data/test_indexed_file.py:480]: should this be commented out?

File: tests/data/test_indexed_file.py [hunk 7]
```diff
@@ -461,6 +465,115 @@ def test_internal_get_gs_signed_url_cache_new_key_if_old_key_expired(
...
+                    user_id=1,
+                    username="some user",
+                    r_pays_project=None,
+                )
+
+                after_cache = db_session.query(AssumeRoleCacheGCP).first()
+
+                assert after_cache
+                assert (
+                    str(type(after_cache))
```
Comment (Avantol13) [inline tests/data/test_indexed_file.py:537]: I'm not sure this assert is worth having. The important thing is that we can json load properly, this type check seems unecessary

File: tests/data/test_indexed_file.py [hunk 8]
```diff
@@ -461,6 +465,115 @@ def test_internal_get_gs_signed_url_cache_new_key_if_old_key_expired(
...
+                    == "<class 'fence.models.AssumeRoleCacheGCP'>"
+                )
+                # check if json loads can properly parse json string stored in cache
+                assert (
+                    str(type(json.loads(after_cache.gcp_private_key)))
```
Comment (Avantol13) [inline tests/data/test_indexed_file.py:542]: I don't think we need this type check either. The assert below will make sure the json.loads works and creates a comparable object to the mock we provided. So I'd remove this line

File: tests/data/test_indexed_file.py [hunk 9]
```diff
@@ -461,6 +465,115 @@ def test_internal_get_gs_signed_url_cache_new_key_if_old_key_expired(
...
+                    == "<class 'dict'>"
+                )
+                assert json.loads(after_cache.gcp_private_key) == sa_private_key
+
+                db_session.delete(after_cache)
+                cleared_cache = db_session.query(AssumeRoleCacheGCP).first()
```
Comment (Avantol13) [inline tests/data/test_indexed_file.py:548]: This isn't the regression test we want, it's not bad to keep this too, but the test we need is:

* Generate url
* Simulate pod rolling (e.g. the **in memory** cache being cleared, **not** the database one)
* Generate url, ensure **new** in memory cache has valid info from the db

File: tests/data/test_indexed_file.py [hunk 10]
```diff
@@ -461,6 +465,115 @@ def test_internal_get_gs_signed_url_cache_new_key_if_old_key_expired(
...
+                    user_id=1,
+                    username="some user",
+                    r_pays_project=None,
+                )
+
+                redo_cache = db_session.query(AssumeRoleCacheGCP).first()
+
+                assert redo_cache
+                assert (
+                    str(type(redo_cache)) == "<class 'fence.models.AssumeRoleCacheGCP'>"
```
Comment (Avantol13) [inline tests/data/test_indexed_file.py:565]: see comments from above. I think we should remove all these type checks and only leave the json.loads assert. And we may want to do something like this to capture the failure correction:

```
try:
  assert json.loads(redo_cache.gcp_private_key) == sa_private_key
except Exception:
  pytest.fail("Could not json.loads(cache)")
```

File: tests/data/test_indexed_file.py [hunk 11]
```diff
@@ -461,6 +465,107 @@ def test_internal_get_gs_signed_url_cache_new_key_if_old_key_expired(
...
+                google_object._generate_google_storage_signed_url(
+                    http_verb="GET",
+                    resource_path="gs://some/location",
+                    expires_in=0,
+                    user_id=1,
+                    username="some user",
+                    r_pays_project=None,
+                )
+
+                try:
```
Comment (Avantol13) [inline tests/data/test_indexed_file.py:536]: you only need the try catch around the json.loads

Comment (Avantol13) [reply tests/data/test_indexed_file.py:536]: so here you can just 
```
assert "1" in google_object._assume_role_cache_gs
assert len(google_object._assume_role_cache_gs["1"]) > 1
assert google_object._assume_role_cache_gs["1"][0] == sa_private_key
```

Comment (Avantol13) [reply tests/data/test_indexed_file.py:536]: and remove the try/except

Comment (BinamB) [reply tests/data/test_indexed_file.py]: Oh, saw these comments after i had added the try-catch for the json.loads. Should i still do this? My reasoning was, whatever is stored in the database should be json.loads-able.

File: tests/data/test_indexed_file.py [hunk 12]
```diff
@@ -461,6 +465,107 @@ def test_internal_get_gs_signed_url_cache_new_key_if_old_key_expired(
...
+                google_object._generate_google_storage_signed_url(
+                    http_verb="GET",
+                    resource_path="gs://some/location",
+                    expires_in=0,
+                    user_id=1,
+                    username="some user",
+                    r_pays_project=None,
+                )
+
+                assert google_object._assume_role_cache_gs["1"][0] == sa_private_key
```
Comment (Avantol13) [inline tests/data/test_indexed_file.py:558]:
```suggestion
```

this is already tested below

**Review by BinamB — APPROVED** [review]

**Pull request merged**